### DISCUSSION TITLE
Mandatory return at the end of functions, even though it might not be reachable?

### DISCUSSION THREAD

**Discussion by FObersteiner** [vlang/v#17591] [Questions and Answers] [answered]
I'm not sure if this is intended behavior, but it was unexpected for me: If I write a forever loop in a function that does specify a return value, it is mandatory to have a return at the end of the function - even though that might not be reachable. Here's a MRE:
```
module main

import rand

fn new_random_key(m map[int]bool) int {
	for {
		key := rand.int()
		if key !in m {
			return key
		}
	}
	// code here is unreachable
}

fn main() {
	mut m := { 1: true }
	// generate a new random key that is not yet in m
	key := new_random_key(m)
	println(key)
}
```
Given a map, the `new_random_key` function should only return a new key if that key is not already in the map. The example might be a bit far-fetched, or could be changed to use return after the loop but you get the point I think ;-) This code won't compile until I put a `return` after the forever-loop. Why?

**Comment by bogen85** [comment]
Good question.

There is a no return attribute `[noreturn]` that can go before the function.
However, that assumes **`no return`** so the return inside the loop would not be accepted.

In your case it is not an actual endless loop, it does have an exit condition.

**Comment by bogen85** [comment, marked as answer]
If you look at https://github.com/vlang/v/blob/6f7192359a9227bdd13ad167838fc43a321c5105/vlib/v/gen/wasm/mem.v#L19 you will see that `panic('unreachable')` is used for situations like this in the V source code.
The idea is that since `panic` is `[noreturn]` you won't get a compile time error in your code when you put in `panic('unreachable')` and don't return anything.

**Reply by FObersteiner** [reply]
that seems like a clear way of saying "something's seriously wrong if you get here".
